src/Avalonia.Controls/Primitives/Popup.cs

```diff
@@ -218,57 +226,71 @@ public bool Topmost
         /// </summary>
         public void Open()
         {
-            if (_popupRoot == null)
+            // Popup is currently open
+            if (_topLevel != null)
+                return;
+            CloseCurrent();
+            var placementTarget = PlacementTarget ?? this.GetLogicalAncestors().OfType<IVisual>().FirstOrDefault();
+            if (placementTarget == null)
+                throw new InvalidOperationException("Popup has no logical parent and PlacementTarget is null");
+            
+            _topLevel = placementTarget.GetVisualRoot() as TopLevel;
+
+            if (_topLevel == null)
             {
-                _popupRoot = new PopupRoot(DependencyResolver)
-                {
-                    [~ContentControl.ContentProperty] = this[~ChildProperty],
-                    [~WidthProperty] = this[~WidthProperty],
-                    [~HeightProperty] = this[~HeightProperty],
-                    [~MinWidthProperty] = this[~MinWidthProperty],
-                    [~MaxWidthProperty] = this[~MaxWidthProperty],
-                    [~MinHeightProperty] = this[~MinHeightProperty],
-                    [~MaxHeightProperty] = this[~MaxHeightProperty],
-                };
-
-                ((ISetLogicalParent)_popupRoot).SetParent(this);
+                throw new InvalidOperationException(
+                    "Attempted to open a popup not attached to a TopLevel");
             }
 
-            _popupRoot.Position = GetPosition();
 
-            if (_topLevel == null && PlacementTarget != null)
+            _popupRoot = new PopupRoot(_topLevel, DependencyResolver)
             {
-                _topLevel = PlacementTarget.GetSelfAndLogicalAncestors().First(x => x is TopLevel) as TopLevel;
+                [~WidthProperty] = this[~WidthProperty],
```

Comment:
We probably want to rethink binding the popup root's sizing parameters to `Popup`: see https://github.com/AvaloniaUI/Avalonia/issues/2373

Doesn't need to be done in this PR, but thought I'd make you aware of that issue.

Reply:
We probably need to make `Popup` to not be a part of the visual tree. Then we could derive from StyledElement directly and define new Width/Height properties